### PR TITLE
Build tools rework

### PR DESCRIPTION
**Pull request by AngheloAlf**
Rework/rewrites various buildtools programs to improve them. Also removes some unused buildtools. This also changes/adds some python dependencies

## Python dependencies:

- Use `crunch64` over `libyaz0`: The former provides both a matching compressor and decompressor (`libyaz0` can't produce matching compression). `crunch64` is magnitudes of time faster than `libyaz0`, the main reason for it is the latter is a Python pure implementation while the former is a Python extension library.
- Use `ipl3checksum` over just having the checksum algorithm there: We either don't know who's the author of the checksum algorithm we are using or it is questionable license-wise. `ipl3checksum` is also usually faster due to it being a Python extension library.

### fixbaserom.py

Replace the checksum algorithm with `ipl3checksum` and replace `libyaz0` with `crunch64`

Runtime: Measured with `time ./tools/fixbaserom.py`

  - `main` branch: `0m8.542s`
  - This PR: `0m0.597s`

93.01% runtime decrease

### decompress_yars.py

Replace `libyaz0` with `crunch64`

Runtime: Measured with `time ./tools/decompress_yars.py`

  - `main` branch: `0m0.296s`
  - This PR: `0m0.093s`

68.58% runtime decrease

### makeyar

Rewrites the original `makeyar` program in Python

Runtime: Measured with the instructions in the following hidden block

<details>
<summary>
Measurement instructions
</summary>
The game contains 7 yar archives that need to be built.

Instead of trying to decide which one to use as a measuring point I decided to measure the time for generating all of them at the same time.

To do this I made a `makeyar_c.sh` script and ran `time` on it. Said file had the following contents:
```bash
tools/buildtools/makeyar build/assets/archives/item_name_static/item_name_static.o build/assets/archives/item_name_static/item_name_static.yar.bin build/assets/archives/item_name_static/item_name_static.symbols.o
tools/buildtools/makeyar build/assets/archives/map_name_static/map_name_static.o build/assets/archives/map_name_static/map_name_static.yar.bin build/assets/archives/map_name_static/map_name_static.symbols.o
tools/buildtools/makeyar build/assets/archives/icon_item_24_static/icon_item_24_static_yar.o build/assets/archives/icon_item_24_static/icon_item_24_static_yar.yar.bin build/assets/archives/icon_item_24_static/icon_item_24_static_yar.symbols.o
tools/buildtools/makeyar build/assets/archives/schedule_dma_static/schedule_dma_static_yar.o build/assets/archives/schedule_dma_static/schedule_dma_static_yar.yar.bin build/assets/archives/schedule_dma_static/schedule_dma_static_yar.symbols.o
tools/buildtools/makeyar build/assets/c/archives/map_i_static/map_i_static.o build/assets/c/archives/map_i_static/map_i_static.yar.bin build/assets/c/archives/map_i_static/map_i_static.symbols.o
tools/buildtools/makeyar build/assets/c/archives/map_grand_static/map_grand_static.o build/assets/c/archives/map_grand_static/map_grand_static.yar.bin build/assets/c/archives/map_grand_static/map_grand_static.symbols.o
tools/buildtools/makeyar build/assets/archives/icon_item_static/icon_item_static_yar.o build/assets/archives/icon_item_static/icon_item_static_yar.yar.bin build/assets/archives/icon_item_static/icon_item_static_yar.symbols.o
```

To measure the Python version I made a `makeyar_py.sh` script that had the same contents but replacing `tools/buildtools/makeyar` with `python3 tools/buildtools/makeyar.py`
</details>

  - `main` branch: `0m1.159s`
  - This PR: `0m1.400s`

20.79% time **increase**
From the whole PR, this is the only change that makes one of the tools slower instead of faster. Even though I think it is worth it because it chops almost 100 lines and due to the decreased complexity of the program. For example the original `makeyar.c` program has a memory leak on the `inputBuf` variable that went unnoticed for many months and which I only noted because of this cleanup.
If we end up preferring the C version instead then I can just undo this.

### elf2rom

Completely removed in favor of `objcopy` and `checksummer.py` (a tool that updates the header checksum and pads the rom appropriately).

Since the rom is generated from the elf by two programs instead of just one it produces the side effect of generating an intermediary `mm.us.rev1.rom_uncompressed.bin` file, which has the rom contents but is not padded correctly nor has a correct checksum

Runtime: Measured with the instructions in the following hidden block

<details>
<summary>
Measurement instructions
</summary>

- To measure `elf2rom`:
    ```bash
    time tools/buildtools/elf2rom -cic 6105 mm.us.rev1.rom_uncompressed.elf mm.us.rev1.rom_uncompressed.z64
    ```
- To measure the new way:
    ```bash
    time (mips-linux-gnu-objcopy --gap-fill=0x00 -O binary mm.us.rev1.rom_uncompressed.elf mm.us.rev1.rom_uncompressed.bin && python3 tools/buildtools/checksummer.py mm.us.rev1.rom_uncompressed.bin mm.us.rev1.rom_uncompressed.z64)
    ```

</details>

  - `main` branch: `0m3.158s`
  - This PR: `0m0.303s`

90.40% time decrease

## Deleted files

- `elf2rom.c`: Superseded by `objcopy` & `checksummer.py`
- `makeyar.c`: Superseded by `makeyar.py`
- `yaz0tool.c`: Unused
- `makeromfs.c`: Unused. I can't really tell what it was its original purpose.
- `n64chksum.c/.h`: Every program depending on this was removed.
- `elf32.c/.h`:Every program depending on this was removed.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -116,13 +116,12 @@ else
...
 
 CPP        := cpp
-ELF2ROM    := tools/buildtools/elf2rom
 MKLDSCRIPT := tools/buildtools/mkldscript
 MKDMADATA  := tools/buildtools/mkdmadata
-YAZ0       := tools/buildtools/yaz0
 ZAPD       := tools/ZAPD/ZAPD.out
 FADO       := tools/fado/fado.elf
-MAKEYAR    := tools/buildtools/makeyar
+MAKEYAR    := python3 tools/buildtools/makeyar.py
```
Comment (hensldm) [inline Makefile:123]: Perhaps better in a different PR, but since you are adding some new python scripts, could you make a `PYTHON` variable (like you do in oot) `PYTHON ?= python3`. Would probably put it in defaults above since I think we want it over writable.

Side note; Would you also be able to fix the `#MIPS toolchain` typo and add a new line between the defaults and `#### Setup ####`

File: tools/buildtools/checksummer.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    rom_bytes.extend([0x00] * (fill_00 - rom_len))
+    rom_bytes.extend([0xFF] * (fill_FF - fill_00))
+
+
+def update_checksum(rom_bytes: bytearray):
+    # Detect CIC
+    # cicKind = ipl3checksum.detectCIC(rom_bytes)
+    # if cicKind is None:
+    #     eprint("Not able to detect CIC, defaulting to 6105")
+    #     cicKind = ipl3checksum.CICKind.CIC_X105
```
Comment (hensldm) [inline tools/buildtools/checksummer.py:33]: Dead cod? Since we don't detect it would it better to be configurable as a function arg with an optional cli arg?

**Review by hensldm — APPROVED** [review]

File: Makefile [hunk 3]
```diff
@@ -116,13 +119,12 @@ else
...
 CPP        := cpp
-ELF2ROM    := tools/buildtools/elf2rom
 MKLDSCRIPT := tools/buildtools/mkldscript
 MKDMADATA  := tools/buildtools/mkdmadata
-YAZ0       := tools/buildtools/yaz0
 ZAPD       := tools/ZAPD/ZAPD.out
 FADO       := tools/fado/fado.elf
-MAKEYAR    := tools/buildtools/makeyar
+MAKEYAR    := $(PYTHON) tools/buildtools/makeyar.py
+CHECKSUMMER:= $(PYTHON) tools/buildtools/checksummer.py
```
Comment (jpburnett) [inline Makefile:127]: I prefer to check winter :)

File: tools/buildtools/makeyar.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+def align_16(val: int) -> int:
+    return (val + 0xF) & ~0xF
+
+
+def create_archive(
+    uncompressed_data: bytearray, symbol_list: list[Symbol]
+) -> bytearray:
+    archive = bytearray()
+
+    first_entry_offset = (len(symbol_list) + 1) * 4
```
Comment (jpburnett) [inline tools/buildtools/makeyar.py:93]:
```suggestion
    first_entry_offset = (len(symbol_list) + 1) << 2
```
Maybe this isn't the python way to do it, and this is my rust/C/C++ brain thinking about it too much...
But in general a shift would be faster than multiplying

**Review by jpburnett — APPROVED** [review]
Nice improvements. That crunch library is pretty nifty, I wonder who helped write it ;)

Left one comment on shifting vs multiplying, not sure if it would really do much but I'd be interested to hear your thoughts.

File: tools/buildtools/makeyar.py (hunk 4 above)
Comment (Dragorn421) [reply tools/buildtools/makeyar.py:93]: Compared to C where a add/shift would translate to one instruction or less, Python has a ton of overhead that makes both options perform about the same

Also in C a good compiler would turn a `*4` into a `<<2` for you anyway!

**Comment by Dragorn421** [comment]
Just curious, could crunch64 replace z64compress, or why not?

File: tools/buildtools/makeyar.py (hunk 4 above)
Comment (AngheloAlf) [reply tools/buildtools/makeyar.py:93]: Seems like Python's multiplication operation is faster than its shift operations
https://stackoverflow.com/questions/54047100/why-are-bitwise-operators-slower-than-multiplication-division-modulo
https://stackoverflow.com/questions/37053379/times-two-faster-than-bit-shift-for-python-3-x-integers

**Comment by AngheloAlf** [comment]
> Just curious, could crunch64 replace z64compress, or why not?

crunch64 only provides the "codec" algorithm (ie yaz0, yay0 and MIO0), so it couldn't replace z64compress by itself but we could write a tool that replaces it by using crunch64.

Why not? Because:

- Said tool would need to be written in Python, which inherently has a slower runtime than a pure C program. I think it would be hard to beat the tool that claims to be "the fastest Zelda 64 rom compressor". The only way I can see this not being an issue is by writing said tool in Rust (since crunch64 is written in Rust), but I don't see people wanting to add yet another language as a dependency
- z64compress besides offering its matching compressor, it also offers a non-matching yaz0 compressor with a better compression rate, which I assume should be useful for modding.
- z64compress offers compressing zelda64 roms with other encodings besides yaz0, like DEFLATE among others, which should beat yaz0 in compression rate at the expense of load times. This may be desirable by modders.


tl;dr: Feasible but probably not worth it

**Comment by Dragorn421** [comment]
I am going to do research, just posting this since I wrote it anyway but I'm not saying anything interesting

-------------

> Said tool would need to be written in Python, which inherently has a slower runtime than a pure C program. I think it would be hard to beat the tool that claims to be "the fastest Zelda 64 rom compressor". The only way I can see this not being an issue is by writing said tool in Rust (since crunch64 is written in Rust), but I don't see people wanting to add yet another language as a dependency

fair, though I don't see a python program adding much overhead over how slow the compression itself is. I guess I could try, doesn't sound that hard?

> z64compress besides offering its matching compressor, it also offers a non-matching yaz0 compressor with a better compression rate, which I assume should be useful for modding.

I should set up a benchmark, but if a modder cares about compression they probably won't use yaz0 anyway

> z64compress offers compressing zelda64 roms with other encodings besides yaz0, like DEFLATE among others, which should beat yaz0 in compression rate at the expense of load times. This may be desirable by modders.

I don't think modding is a good enough reason to keep a tool around, but I understand it's a bonus in addition to the first point

**Comment by AngheloAlf** [comment]
I totally think it is worth a shot, at least to know how faster/slower that tool can be compared to z64compress.

Personally I'm not that interested on the modding benefits, but I see the appeal and I think having a modding friendly repo is nice to have either way

File: Makefile (hunk 3 above)
Comment (hensldm) [reply Makefile:127]:
```suggestion
CHECKSUMMER := $(PYTHON) tools/buildtools/checksummer.py
```
Could use another space. Might want to add one to the other tools as well to keep them aligned.

**Comment by hensldm** [comment]
> Since the rom is generated from the elf by two programs instead of just one it produces the side effect of generating an intermediary mm.us.rev1.rom_uncompressed.bin file, which has the rom contents but is not padded correctly nor has a correct checksum

Just realized, running a `make clean` should also remove this bin file, which it does not currently do. Or I wonder if we should just remove it right after `checksummer` runs?

**Comment by Dragorn421** [comment]
> > Since the rom is generated from the elf by two programs instead of just one it produces the side effect of generating an intermediary mm.us.rev1.rom_uncompressed.bin file, which has the rom contents but is not padded correctly nor has a correct checksum
> 
> Just realized, running a `make clean` should also remove this bin file, which it does not currently do. Or I wonder if we should just remove it right after `checksummer` runs? Another option would be just have checksummer write to the same file it reads from?

Imo they shouldn't be separate files at all for performance reasons, it's costly and wasteful to write two ~50MiB files instead of only one that gets modified

**Comment by AngheloAlf** [comment]
Reimplemented checksummer to modify the input rom instead of producing a new one.

Old:
```
$ time ./tools/buildtools/checksummer.py mm.us.rev1.rom_uncompressed.bin mm.us.rev1.rom_uncompressed.z64

real    0m0.184s
user    0m0.061s
sys     0m0.122s
```

New:
```
$ time ./tools/buildtools/checksummer_file.py mm.us.rev1.rom_uncompressed.bin 

real    0m0.078s
user    0m0.051s
sys     0m0.026s
```

**Comment by ginomeelectra202** [comment]
?emiT fo aniracO ni si ereht ekil ti esaercni ot yaw a ereht si dna pac eht si tahw ,os fI ?emit eno yna ta dloh nac uoy stuN ukeD fo rebmun mumixam a ereht sI

**Comment by hirottau302** [comment]
so I own a game shark version 3.0 and I have been having a lot of trouble with it. so for some games it works just fine, but for most of my games, majora's mask, daddy Kong racing, DK 64, ocarina of time when i boot up the game with my cheats enabled, the thing just shuts off. this happens every single time even if there are no cheats enabled it just shuts down. I think it have something to do with the enable code some games require. does anyone have similar troubles and have an answer?

**Comment by ginomeelectra202** [comment]
?taht pots I od woh ,ti gniknird pu dne I ,noitop der eht emuoK evig ot yrt I revenehw

**Comment by altoidmerope294** [comment]
I pick up a bomb, wait until the 3rd red flash, drop it by shielding, step back until link cannot pick it up, roll into it, then press the pick up button to pick it up and it explodes against my shield at the last second (at least I think so), but nothing happens except the bombs, including other all c items, become unusable until I enter a new area. Is it me or the system? Thanks

**Comment by ginomeelectra202** [comment]
I'm looking for her but I can't find her. No real reason but I just wanna know where she went. Can someone help?

**Comment by altoidmerope294** [comment]
.?ssob eht dnif tnac i tub MOOR ssob eht ot tog evah I

**Comment by ginomeelectra202** [comment]
?krow tuctrohs eht ekam ot gnissim ma I gnihtemos ereht si ,niaga parc llew/ymmum elohw taht od ot tnaw TON OD yllaer I .ecnereffid a edam rehtien tub elosnoc eht gnitratser neve dna tsrif gnirps eht morf gniwolf retaw eht gnitteg deirt ev'I !raeppasid t'now kcolb eht tub ,thgil eht tcelfer ot dleihs rorrim eht esu nac I .moor taht sretne thgil dna latsyrc eht tih I

!krow t'nseod tuctrohs eht tub dleihs rorrim eht evah llits I ,no teg I nehw woN .ffo tog dna emag eht devas I dekrow tuctrohs ehT .raeppasid ti ekam ot kcalb eht otno thgil eht gninihs dna ,latsyrc eht ,etag niam eht ot txen llaw eht ni eloh eht gnivlovni tuctrohs taht tuo kcehc ot erus edam dna eltsac anakI derolpxe ylfeirb I

.dleihs rorrim eht deviecer I dna )ssa eht ni niap( gnidart dna seimmum eht lla htiw noitces lleW eht htaeneB eritne taht detelpmoc tsuj I

**Comment by hirottau302** [comment]
I don't intend to play the Song of Time. I just want to buy some Green Potion and come back. Will I lose all of my work?

**Comment by yemenalcyone230** [comment]
I need help

**Review by engineer124 — APPROVED** [review]
